extension-manifest-v3/src/background/adblocker.js

```diff
@@ -24,102 +25,116 @@ import { getMetadata } from '/utils/trackerdb.js';
+        await Promise.all(
+          enabledEngines.map((id) =>
+            engines.init(id).catch(() => {
+              console.error(`[adblocker] failed to load engine: ${id}`);
+              return null;
+            }),
+          ),
+        )
+      ).filter((engine) => engine),
+    );
 
-  // Experimental filters
-  observe('experimentalFilters', async (value, lastValue) => {
-    engines.setEnv('env_experimental', value);
+    console.info(
+      `[adblocker] engine reloaded with: ${enabledEngines.join(', ')}`,
+    );
+  } else {
+    engines.create(engines.MAIN_ENGINE);
+    console.info('[adblocker] Engine reloaded with no filters');
+  }
+}
 
-    // As engines on the server might have new filters, we force the update
-    // when value has changed from false to true
-    if (lastValue !== undefined && value) {
-      engines.updateAll().catch(() => null);
-    }
-  }),
+engines.addChangeListener(engines.CUSTOM_ENGINE, async () => {
+  reloadEngine(getEnabledEngines(options));
+});
 
-  // Regional filters
-  observe('regionalFilters', async ({ enabled, regions }, lastValue) => {
-    // Background startup
-    if (!lastValue) {
-      const engine = await regionalFiltersEngine;
-      // Pre-requirement for skipping update - engine must be initialized
-      // Otherwise it is a very first try to setup the engine
-      if (engine.lists.size) return;
-    }
+export const setup = asyncSetup([
+  observe(async (value, lastValue) => {
+    options = value;
 
-    // Clean previous regional engines
-    if (lastValue) {
-      lastValue.regions
-        .filter((id) => !regions.includes(id))
-        .forEach((id) => engines.clean(`lang-${id}`));
+    const enabledEngines = getEnabledEngines(value);
+    const prevEnabledEngines = lastValue && getEnabledEngines(lastValue);
+
+    if (
+      // Reload/mismatched main engine
+      !(await engines.init(engines.MAIN_ENGINE)) ||
+      // Enabled engines changed
+      (prevEnabledEngines &&
+        (enabledEngines.length !== prevEnabledEngines.length ||
+          enabledEngines.some((id, i) => id !== prevEnabledEngines[i]))) ||
+      // Engine filters updatedAt changed after successful update
+      (lastValue && value.filtersUpdatedAt > lastValue.filtersUpdatedAt)
+    ) {
+      // The regional filters engine is no longer used, so we must remove it
+      // from the storage. We do it as rarely as possible, to avoid unnecessary loads.
+      // TODO: this can be removed in the future release when most of the users will have
+      // the new version of the extension
+      engines.remove('regional-filters');
+
+      await reloadEngine(enabledEngines);
     }
 
-    // Schedule merge when one of the regional engines is updated
-    regions.forEach((id) => {
-      engines.addUpdateListener(`lang-${id}`, mergeRegionalEngines);
-    });
+    if (
+      // Experimental filters enabled
+      (options.experimentalFilters &&
+        lastValue?.experimentalFilters === false) ||
```

Comment:
If lastValue does not exist, this check will always fail. Is that expected?

Reply:
Yes, as this should only pass through is `lastValue` is set.

In simple words the condition is: "When a user actively changed the setting to `true`"